### PR TITLE
add user passthrough column to users model and update dependencies

### PR DESCRIPTION
**Pull request by fivetran-reneeli**
## PR Overview
**This PR will address the following Issue/Feature:**
https://github.com/fivetran/dbt_iterable/issues/43
and internal ticket

**This PR will result in the following new package version:** v0.12.0

Non breaking as the default config for event_extension is still true so existing users won't see any changes. And the passthrough columns only add columns, instead of deleting or modifying existing ones.

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

## Bug Fixes
- Introduces variable `iterable__using_event_extension` to disable the `event_extension` table and exclude its fields from persisting downstream. This allows the downstream models to run even if the source `event_extension` table does not exist. For more information on how to configure the `iterable__using_event_extension` variable, refer to the [README](https://github.com/fivetran/dbt_iterable/blob/main/README.md#step-4-enablingdisabling-models).
- Persists `user_history` passthrough columns as stipulated via the `iterable_user_history_pass_through_columns` variable through to the `iterable__users` model. For more information on how to configure the `iterable_user_history_pass_through_columns` variable, refer to the [README](https://github.com/fivetran/dbt_iterable/blob/main/README.md#passing-through-additional-fields).

## Under the Hood
- Addition of integrity and consistency validation tests within integration tests pertaining to the `iterable__user_unsubscriptions` model.
- Updated seed data to ensure proper testing of the latest [v0.8.1 `dbt_iterable_source` release](https://github.com/fivetran/dbt_iterable_source/releases/tag/v0.8.1).
- Updated [pull request and issue templates](https://github.com/fivetran/dbt_iterable_source/tree/v0.8.1/.github).
- Included auto-releaser GitHub Actions workflow to automate future releases.

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [x] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
- [x] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:

Variables tried:

```
    iterable_user_history_pass_through_columns:
      - name: phone_number_line_type
        alias: phone_digits
      - name: additional_properties
    iterable_event_extension_pass_through_columns:
      - name: browser_token
    iterable__using_event_extension: true
#    iterable__using_event_extension: false
    
```
### If you had to summarize this PR in an emoji, which would it be?
 
:dancer:

### CODE REVIEW

**Comment by fivetran-reneeli** [comment]
will regen docs upon approval

File: .buildkite/scripts/run_models.sh
Comment (fivetran-avinash) [inline .buildkite/scripts/run_models.sh:25]: Should there be an incremental run here as well, since `iterable__using_event_extension` is being run on `iterable__events`, an incremental model?

File: CHANGELOG.md
Comment (fivetran-avinash) [inline CHANGELOG.md:1]: There are changes to when `experiment_id` is or is not being brought in this model, plus the new aggregate metrics, which I believe classifies as a schema change. Should we consider making this a breaking change and bumping the version up?

File: CHANGELOG.md
Comment (fivetran-avinash) [inline CHANGELOG.md:8]: You can add more details around the consistency tests you've created here too.

File: README.md
Comment (fivetran-avinash) [inline README.md:110]:
```suggestion
This package includes fields we judged were standard across Iterable users. However, the Fivetran connector allows for additional columns to be brought through in the `event_extension` and `user_history` objects. Therefore, if you wish to bring them through, leverage our passthrough column variables. For `event_extension` columns, ensure that `iterable__using_event_extension` is set to True, which is the default.
```

File: README.md
Comment (fivetran-avinash) [inline README.md:112]: Is this meant to be `iterable__user_campaign`?

File: integration_tests/dbt_project.yml
Comment (fivetran-avinash) [inline integration_tests/dbt_project.yml:25]: I think you can uncomment this, Buildkite is equipped to handle it.

File: integration_tests/dbt_project.yml
Comment (fivetran-avinash) [inline integration_tests/dbt_project.yml:6]: Can we switch the `iterable_schema` to be under `iterable_source` as it was before, since the models are supposed to be referencing `iterable_source.<iterable_schema>` I think when doing seed/run? Or are there advantages to this approach?

File: integration_tests/tests/consistency/consistency_users.sql
Comment (fivetran-avinash) [inline integration_tests/tests/consistency/consistency_users.sql:9]: When trying to run this test I received this error.

```
10:02:09    Database Error in test consistency_users (tests/consistency/consistency_users.sql)
  Unrecognized name: created_at; Did you mean updated_at? at [12:16]
  compiled Code at target/run/iterable_integration_tests/tests/consistency/consistency_users.sql

```

When looking at `iterable__users`, looks like `created_at` should be changed to `updated_at`.

File: models/iterable__users.sql
Comment (fivetran-avinash) [inline models/iterable__users.sql:1]: What a cool macro! Really like the flexibility here.

File: packages.yml
Comment (fivetran-avinash) [inline packages.yml:5]: Reminder to switch the package version back before release.

File: CHANGELOG.md
Comment (fivetran-avinash) [inline CHANGELOG.md:3]: I'd add more about the `experiment_id` updates (plus the new columns you're adding) in either a Breaking Change section, or in Bug Fixes below, whichever makes more sense.  It looks like it's mostly related to the `iterable__using_event_extension` section, but more context would be helpful since the logic introduced is non-intuitive.

File: README.md
Comment (fivetran-avinash) [inline README.md:1]: @fivetran-reneeli You will need to add the var `iterable__using_event_extension` to the yml section in step 4.

File: CHANGELOG.md
Comment (fivetran-avinash) [inline CHANGELOG.md:6]: This link--https://github.com/fivetran/dbt_iterable/blob/main/README.md#passing-through-additional-fields--does not work as it's in a collapsed section of the README. I'd remove lines 105-106 and test if the hyperlink works on your branch then.

File: models/iterable__events.sql
Comment (fivetran-avinash) [inline models/iterable__events.sql:69]:
```suggestion
        {{ dbt_utils.star(from=ref('stg_iterable__event_extension'), except=exclude_fields) }}
```

File: models/intermediate/int_iterable__campaign_event_metrics.sql
Comment (fivetran-avinash) [inline models/intermediate/int_iterable__campaign_event_metrics.sql:22]: @fivetran-reneeli 

Two questions:  
1) Is there a reason we are summing `experiment_id`? (experiment ids usually don't have much aggregate value in my experience). The count logic makes sense. I think adding details around both new columns to the CHANGELOG would be helpful, since it looks as if they are being brought into end models.

2) `experiment_id` would be a duplicate column if this logic compiled (since `iterable__using_event_extension` being True holds for both cases when `experiment_id` is brought in by line 12). I think this column value needs to be changed because if it compiles it will lead to duplicate errors.

3) These changes will impact `iterable__campaigns`. I'm wondering if we should also write a similar consistency test for that model as well to be on the safe side.

File: integration_tests/tests/consistency/consistency_events.sql
Comment (fivetran-avinash) [inline integration_tests/tests/consistency/consistency_events.sql:1]: I got this error when trying to test this model in the integration_tests validation run:

```
10:02:09    Database Error in test consistency_events (tests/consistency/consistency_events.sql)
  Queries in EXCEPT DISTINCT have mismatched column count; query 1 has 73 columns, query 2 has 72 columns at [25:5]
```

This is likely because `experiment_id` now is being removed in the default version. It shouldn't be too difficult to test that it successfully ran though--can you try and document your process for a successful test here?

File: models/intermediate/int_iterable__campaign_event_metrics.sql
Comment (fivetran-avinash) [inline models/intermediate/int_iterable__campaign_event_metrics.sql:19]: @fivetran-reneeli Were you able to successfully test that this logic worked downstream in bringing in `experiment_id`? I had trouble finding a way for this logic to work, because I had a hypothesis that this would not compile because of the potential duplicate column issue. Could you add a successful run of this logic to your validation notes?

File: models/intermediate/int_iterable__user_event_metrics.sql
Comment (fivetran-avinash) [inline models/intermediate/int_iterable__user_event_metrics.sql:21]: Similar note to `int_iterable__campaign_events`.

File: models/intermediate/int_iterable__user_event_metrics.sql
Comment (fivetran-avinash) [inline models/intermediate/int_iterable__user_event_metrics.sql:24]: Similar note to `int_iterable__campaign_events`.

File: models/intermediate/int_iterable__campaign_event_metrics.sql
Comment (fivetran-avinash) [inline models/intermediate/int_iterable__campaign_event_metrics.sql:5]:
```suggestion
{%- set user_campaign_columns = adapter.get_columns_in_relation(ref('iterable__user_campaign')) %}
```

File: integration_tests/tests/consistency/consistency_users.sql
Comment (fivetran-avinash) [inline integration_tests/tests/consistency/consistency_users.sql:15]:
```suggestion
    where date(updated_at) < date({{ dbt.current_timestamp() }})
```

File: models/intermediate/int_iterable__campaign_event_metrics.sql
Comment (fivetran-avinash) [inline models/intermediate/int_iterable__campaign_event_metrics.sql:5]:
```suggestion
{%- set user_campaign_columns = adapter.get_columns_in_relation(ref('iterable__user_campaign')) %}
```

**Review by fivetran-avinash — CHANGES_REQUESTED** [review]
@fivetran-reneeli Awesome job! This is a pretty extensive update with a lot of custom logic you had to implement, so great work in getting it to this point.

Some comments to address before approval. Let me know if you have any questions.

File: integration_tests/dbt_project.yml
Comment (fivetran-reneeli) [reply integration_tests/dbt_project.yml:25]: done!

File: models/intermediate/int_iterable__campaign_event_metrics.sql
Comment (fivetran-reneeli) [reply models/intermediate/int_iterable__campaign_event_metrics.sql:19]: Thanks Avinash for noticing this, `experiment_id` was not meant to be part of the aggregations and I see that the logic I added didn't exclude for it when `iterable__using_event_extension` is True. It looks like just keeping it in the exclusion list does the trick.

File: models/intermediate/int_iterable__user_event_metrics.sql
Comment (fivetran-reneeli) [reply models/intermediate/int_iterable__user_event_metrics.sql:24]: updated

File: integration_tests/tests/consistency/consistency_events.sql
Comment (fivetran-reneeli) [reply integration_tests/tests/consistency/consistency_events.sql:1]: Thanks for catching this-- I reran the tests and see the errors now. Actually, `experiment_id` should be there as default so this was unexpected. This led me to looking into the prod/dev runs and actually the missing column is `channel_type`. Hmm. I will continuing looking into why this is happening

image

Comment (fivetran-avinash) [reply integration_tests/tests/consistency/consistency_events.sql:1]: @fivetran-reneeli Strange. Can you look in the upstream models and see if they are present there?

File: README.md
Comment (fivetran-reneeli) [reply README.md:112]: `iterable__user_campaign` doesn't involve any pass through columns I believe.

File: integration_tests/tests/consistency/consistency_events.sql
Comment (fivetran-reneeli) [reply integration_tests/tests/consistency/consistency_events.sql:1]: Thanks Avinash for helping troubleshoot and realize it was due to a missing comma!

File: CHANGELOG.md
Comment (fivetran-reneeli) [reply CHANGELOG.md:3]: Thanks Avinash, I added more info in Under the Hood for `iterable__using_event_extension` and the related `experiment_id` field

File: .buildkite/scripts/run_models.sh
Comment (fivetran-reneeli) [reply .buildkite/scripts/run_models.sh:25]: Do you mind clarifying-- do you mean adding an extra `dbt run` after ?

File: CHANGELOG.md
Comment (fivetran-reneeli) [reply CHANGELOG.md:9]: added!

File: CHANGELOG.md
Comment (fivetran-reneeli) [reply CHANGELOG.md]: good point-- bumped to breaking change

File: integration_tests/dbt_project.yml
Comment (fivetran-reneeli) [reply integration_tests/dbt_project.yml]: switched back!

File: integration_tests/tests/consistency/consistency_users.sql
Comment (fivetran-reneeli) [reply integration_tests/tests/consistency/consistency_users.sql]: Thanks for catching!

File: README.md
Comment (fivetran-reneeli) [reply README.md:1]: added

File: CHANGELOG.md
Comment (fivetran-reneeli) [reply CHANGELOG.md:7]: Removed the collapsed block

File: models/intermediate/int_iterable__campaign_event_metrics.sql
Comment (fivetran-reneeli) [reply models/intermediate/int_iterable__campaign_event_metrics.sql]: Added experiment_id back to the exclusion

File: models/intermediate/int_iterable__user_event_metrics.sql
Comment (fivetran-reneeli) [reply models/intermediate/int_iterable__user_event_metrics.sql]: Added experiment_id back to the exclusion

File: .buildkite/scripts/run_models.sh
Comment (fivetran-avinash) [reply .buildkite/scripts/run_models.sh:25]: @fivetran-reneeli So basically
  
```suggestion
dbt run --vars '{iterable__using_campaign_label_history: false, iterable__using_user_unsubscribed_message_type_history: false, iterable__using_campaign_suppression_list_history: false, iterable__using_event_extension: false}' --target "$db" 
```

File: README.md
Comment (fivetran-avinash) [inline README.md:1]: Reminder to bump the README `dbt_iterable` version in lines 76-80, now that it's a breaking change.

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-reneeli thanks for applying all these changes! I verified that the validation tests worked as expected.

A few comments left to address (plus another reminder to regenerate the docs) and then you should be good for approval.

File: README.md
Comment (fivetran-reneeli) [reply README.md:1]: ah thank you, updated

File: .buildkite/scripts/run_models.sh
Comment (fivetran-reneeli) [reply .buildkite/scripts/run_models.sh:25]: oh gotcha

**Review by fivetran-avinash — APPROVED** [review]
Hi @fivetran-reneeli ! Thanks for addressing all the additional comments! This looks good to go. 

Regarding your script issue and your commit history, it looks like I sent a final incorrect comment. My initial recommendation was to add the additional incremental run script after the full refresh. But my last suggestion in Github did indeed remove it. Sorry, that line should remain unchanegd.

Can you test this configuration in Buildkite that runs a full-refresh, then an incremental run, and see if this works:


```
dbt run --vars '{iterable__using_campaign_label_history: false, iterable__using_user_unsubscribed_message_type_history: false, iterable__using_campaign_suppression_list_history: false, iterable__using_event_extension: false}' --target "$db" —full-refresh

dbt run --vars '{iterable__using_campaign_label_history: false, iterable__using_user_unsubscribed_message_type_history: false, iterable__using_campaign_suppression_list_history: false, iterable__using_event_extension: false}' --target "$db"
```

File: .buildkite/scripts/run_models.sh
Comment (fivetran-avinash) [reply .buildkite/scripts/run_models.sh]: @fivetran-reneeli See PR approval comment.

**Comment by fivetran-reneeli** [comment]
> Hi @fivetran-reneeli ! Thanks for addressing all the additional comments! This looks good to go.
> 
> Regarding your script issue and your commit history, it looks like I sent a final incorrect comment. My initial recommendation was to add the additional incremental run script after the full refresh. But my last suggestion in Github did indeed remove it. Sorry, that line should remain unchanegd.
> 
> Can you test this configuration in Buildkite that runs a full-refresh, then an incremental run, and see if this works:
> 
> ```
> dbt run --vars '{iterable__using_campaign_label_history: false, iterable__using_user_unsubscribed_message_type_history: false, iterable__using_campaign_suppression_list_history: false, iterable__using_event_extension: false}' --target "$db" —full-refresh
> 
> dbt run --vars '{iterable__using_campaign_label_history: false, iterable__using_user_unsubscribed_message_type_history: false, iterable__using_campaign_suppression_list_history: false, iterable__using_event_extension: false}' --target "$db"
> ```

Thanks @fivetran-avinash yeah I had figured this was what you meant and I tried that in [this commit](https://github.com/fivetran/dbt_iterable/pull/44/commits/ce3b63497ecce08781c45c398191eb25602e0e70), while it failed I think it's because I didn't add the vars the second time around. Just pushed this and should be good. Appreciate your review!!

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-reneeli thanks for working through these changes. Please see my comments below to address before this will be ready for release approval.

Additionally, I noticed that `dbt test` fails if you have set `iterable__using_event_extension` to `false`. This is because in the [yml where we define the tests](https://github.com/fivetran/dbt_iterable/blob/a8032b764d21e91f6c9c859a66563fac7d0049ce/models/iterable.yml#L5-L12) we are still including `experiment_id` and not catching if the variable is turned on/off.

image

I recommend it would probably be best to create a surrogate key in the model that defines the uniqueness of the end model (like [here](https://github.com/fivetran/dbt_netsuite/blob/dc306db9ba31d78db9072d78c86d30a9bd04effd/models/netsuite2/netsuite2__balance_sheet.sql#L273-L282) in our netsuite model) and make that surrogate key dynamic so we can then test that field in the yml (like we have done [here](https://github.com/fivetran/dbt_netsuite/blob/dc306db9ba31d78db9072d78c86d30a9bd04effd/models/netsuite2.yml#L104-L108) for the same netsuite model). 

Lastly, I know this is out of scope for this task, but would you be able to create an issue for us to explore the `iterable__events` incremental logic. I did an integrity test on our internal data (albeit our internal data may be incomplete and not emulate a customer environment) and found that the row count for the `iterable__events` model does not match the row count of the `stg_iterable__event` model following an incremental run (this is present in both dev and prod). It would be great for us to proactively create a ticket for us to explore this in a future update.

image

**Comment by fivetran-reneeli** [comment]
> @fivetran-reneeli thanks for working through these changes. Please see my comments below to address before this will be ready for release approval.
> 
> Additionally, I noticed that `dbt test` fails if you have set `iterable__using_event_extension` to `false`. This is because in the [yml where we define the tests](https://github.com/fivetran/dbt_iterable/blob/a8032b764d21e91f6c9c859a66563fac7d0049ce/models/iterable.yml#L5-L12) we are still including `experiment_id` and not catching if the variable is turned on/off.
> 
> image
> 
> I recommend it would probably be best to create a surrogate key in the model that defines the uniqueness of the end model (like [here](https://github.com/fivetran/dbt_netsuite/blob/dc306db9ba31d78db9072d78c86d30a9bd04effd/models/netsuite2/netsuite2__balance_sheet.sql#L273-L282) in our netsuite model) and make that surrogate key dynamic so we can then test that field in the yml (like we have done [here](https://github.com/fivetran/dbt_netsuite/blob/dc306db9ba31d78db9072d78c86d30a9bd04effd/models/netsuite2.yml#L104-L108) for the same netsuite model).
> 
> Lastly, I know this is out of scope for this task, but would you be able to create an issue for us to explore the `iterable__events` incremental logic. I did an integrity test on our internal data (albeit our internal data may be incomplete and not emulate a customer environment) and found that the row count for the `iterable__events` model does not match the row count of the `stg_iterable__event` model following an incremental run (this is present in both dev and prod). It would be great for us to proactively create a ticket for us to explore this in a future update.
> 
> image

Thanks @fivetran-joemarkiewicz  for catching-- I removed experiment_id from anywhere it was included in a unique test and instead created a surrogate key in those models (iterable__user_campaign and iterable__campaigns).

Hmm strange about the row counts-- will make issue for that

File: packages.yml
Comment (fivetran-reneeli) [reply packages.yml]: oops- removed

File: packages.yml
Comment (fivetran-joemarkiewicz) [reply packages.yml:4]: Reminder to switch before merge and release

File: CHANGELOG.md
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:5]: You'll also want to add in here that the uniqueness test has been updated to account for instances when`iterable__using_event_extension` is enabled/disabled. It would be good to add what the new surrogate fields are and which models they can be found in.

**Review by fivetran-joemarkiewicz — APPROVED** [review]
@fivetran-reneeli thanks for working through this PR. I approved, but have a request to add another CHANGELOG entry for the new surrogate key fields and their purpose for the dynamic uniqueness tests. No need to block release approval, but please make sure that comment is addressed and the `packages.yml` is updated before being merged and released. Thanks!

File: CHANGELOG.md
Comment (fivetran-reneeli) [reply CHANGELOG.md:5]: updated!

**Pull request merged**
